### PR TITLE
Sync Javascript Binding - Add support for property interception

### PR DESCRIPTION
**Pull request by santoshmgh**
**Summary:**
   -  I have added property interceptor for Javascript Binding

**Changes:** 
   - I have modified the TryGetProperty and TrySetProperty in JavascriptObjectRepository 
   - I have PropertyInterceptor as the BindingOptions
   - Added IPropertyInterceptor to CefSharp.ModelBinding
   - Added new Unit Test cases for Property Interceptor to  CefSharp.Test.JavascriptBinding.JavaScriptObjectRepositoryFacts
**How Has This Been Tested?**  
  - I have tested it using CefSharp.Wpf.Example 

**Types of changes**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Updated documentation

**Checklist:**

- [x] Tested the code(if applicable)
- [ ] Commented my code
- [ ] Changed the documentation(if applicable)
- [ ] New files have a license disclaimer
- [ ] The formatting is consistent with the project (project supports .editorconfig)

### CODE REVIEW

**Comment by santoshmgh** [comment]
@amaitland Kindly let me know if I am missing anything for the failing test case.

**Comment by amaitland** [comment]
The failing test is #3867 

Property interception is not available in .Net Core and will need to be excluded with an if def see https://github.com/cefsharp/CefSharp/blob/master/CefSharp/Internals/JavascriptObjectRepository.cs#L197 for example.

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    public interface IPropertyInterceptor
+    {
+        /// <summary>
+        /// Intercept the get property
+        /// </summary>
+        /// <param name="property"></param>
+        /// <param name="parameters"></param>
+        /// <param name="propertName"></param>
+        /// <returns></returns>
+        object GetIntercept(Func<object> property, string propertName);
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:22]: Method names Should likely be InterceptGet and InterceptSet respectively.

Xml doc is incomplete, needs to be fixed.

File: CefSharp/BindingOptions.cs [hunk 2]
```diff
@@ -29,5 +29,12 @@ public static BindingOptions DefaultBinder
         /// for logging calls (from js) to .net methods.
         /// </summary>
         public IMethodInterceptor MethodInterceptor { get; set; }
+
+        /// <summary>
+        /// Interceptor used for intercepting calls to the target object property. For instance, can be used
```
Comment (amaitland) [inline CefSharp/BindingOptions.cs:34]: > Interceptor used for intercepting calls to the target object property

Interceptor used for intercepting get/set calls to the target object property

File: CefSharp/ModelBinding/IPropertyInterceptor.cs (hunk 1 above)
Comment (amaitland) [reply CefSharp/ModelBinding/IPropertyInterceptor.cs:22]: Property parameter should be renamed to be more descriptive, possibly propertyGetter

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <summary>
+        /// Intercept the set property
+        /// </summary>
+        /// <param name="property"></param>
+        /// <param name="parameters"></param>
+        /// <param name="propertName"></param>
+        /// <returns></returns>
+        void SetIntercept(Action<Object> property, object parameter, string propertName);
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:31]: Property parameter should be renamed to be more descriptive, possibly propertySetter

File: CefSharp/ModelBinding/IPropertyInterceptor.cs (hunk 1 above)
Comment (santoshmgh) [reply CefSharp/ModelBinding/IPropertyInterceptor.cs]: Incorporated the comments.

File: CefSharp/BindingOptions.cs (hunk 2 above)
Comment (santoshmgh) [reply CefSharp/BindingOptions.cs]: Incorporated the comment.

File: CefSharp/ModelBinding/IPropertyInterceptor.cs (hunk 3 above)
Comment (santoshmgh) [reply CefSharp/ModelBinding/IPropertyInterceptor.cs]: Incorporated the comment

**Comment by santoshmgh** [comment]
@amaitland Are we good to merge these changes?

**Comment by amaitland** [comment]
> Property interception is not available in .Net Core and will need to be excluded with an if def see https://github.com/cefsharp/CefSharp/blob/master/CefSharp/Internals/JavascriptObjectRepository.cs#L197 for example.

This will need to be addressed before this PR can be considered for merge. 

A unit test should also be added. Testing the JavaScript object repository functionality directly should be sufficient. Add test to https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/JavascriptBinding/JavaScriptObjectRepositoryFacts.cs

**Comment by santoshmgh** [comment]
@amaitland Thanks for sharing the above pointer
If my understanding is correct Property interception execution in TryGetProperty and TrySetProperty should be excluded with NETCOREAPP

**Comment by amaitland** [comment]
> Example: : TryGetProperty

To simplify things the whole TryGetProperty method can be excluded from the netcore implementation. 

> Should I need to take care of the same at any other place apart from TryGetProperty and TrySetProperty API?

Yes. You will need to compile the netcore solution and resolve all the errors.

File: CefSharp.Example/ModelBinding/PropertyInterceptorLogger.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// Copyright © 2017 The CefSharp Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+using System.Diagnostics;
+using CefSharp.ModelBinding;
+
+namespace CefSharp.Example.ModelBinding
```
Comment (amaitland) [inline CefSharp.Example/ModelBinding/PropertyInterceptorLogger.cs:9]: This file should be excluded from the `netcore csproj` file.

File: CefSharp/BindingOptions.cs [hunk 5]
```diff
@@ -29,5 +29,12 @@ public static BindingOptions DefaultBinder
         /// for logging calls (from js) to .net methods.
         /// </summary>
         public IMethodInterceptor MethodInterceptor { get; set; }
+
+        /// <summary>
+        /// Interceptor used for intercepting get/set calls to the target object property. For instance, can be used 
+        /// for logging calls to .net property (from js)
+        /// </summary>
+        public IPropertyInterceptor PropertyInterceptor { get; set; }
```
Comment (amaitland) [inline CefSharp/BindingOptions.cs:37]: Should be excluded from `netcore` project, shouldbe something like the following. Same will apply to other references.

```c#
#if !NETCOREAPP
        /// <summary>
        /// Interceptor used for intercepting get/set calls to the target object property. For instance, can be used 
        /// for logging calls to .net property (from js)
        /// </summary>
        public IPropertyInterceptor PropertyInterceptor { get; set; }
#endif
```

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+using System;
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:1]: This file should be excluded from the `netcore csproj` file.

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        /// <summary>
+        /// Called before the get property is invokved. You are now responsible for evaluating
+        /// the property and returning the result.
+        /// </summary>
+        /// <param name="propertyGetter">A Func that represents the property to be called</param>
+        /// <param name="propertName">Name of the property to be called</param>
+        /// <returns>The property result</returns>
+        /// <example>
+        /// 
+        ///public object IPropertyInterceptor.InterceptGet(Func<object> propertyGetter, string propertyName)
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:24]: I doubt this is valid xml doc as it contains angle brackets for the `Func`. 

Either encode the brackets or put the code in a `<code/>` block with `CDATA`

```c#
/// <example>
/// <code>
/// <![CDATA[
/// public object IPropertyInterceptor.InterceptGet(Func<object> propertyGetter, string propertyName)
/// {
///    object result = propertyGetter();
///    Debug.WriteLine("InterceptGet " + propertyName);
///    return result;
/// }
/// ]]>
/// </code>
/// </example>
```

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        /// <summary>
+        /// Called before the set property is invokved. You are now responsible for evaluating
+        /// the property.
+        /// </summary>
+        /// <param name="propertySetter">A Func that represents the property to be called</param>
+        /// <param name="parameter">paramater to be set to property</param>
+        /// <param name="propertName">Name of the property to be called</param>
+        /// <example>
+        /// 
+        /// public void IPropertyInterceptor.InterceptSet(Action<object> propertySetter, object parameter, string propertName)
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:43]: Same as above

File: CefSharp/ModelBinding/IPropertyInterceptor.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        /// {
+        ///    Debug.WriteLine("InterceptSet " + propertName);
+        ///    propertySetter(parameter);
+        /// }
+        /// </example>
+        ///
+        void InterceptSet(Action<Object> propertySetter, object parameter, string propertName);
+    }
+}
+
```
Comment (amaitland) [inline CefSharp/ModelBinding/IPropertyInterceptor.cs:53]: Remove extra lines

File: CefSharp.Example/ModelBinding/PropertyInterceptorLogger.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
+// Copyright © 2017 The CefSharp Authors. All rights reserved.
```
Comment (amaitland) [inline CefSharp.Example/ModelBinding/PropertyInterceptorLogger.cs:1]: Fix copyright year

**Review by amaitland — CHANGES_REQUESTED** [review]
Comments inline

**Comment by amaitland** [comment]
@santoshmgh Please confirm if you are planning on implementing the required changes or if the `PR` should be closed. Thanks.

**Comment by santoshmgh** [comment]
@amaitland I will be incorporating the above comments..

File: CefSharp/BindingOptions.cs (hunk 5 above)
Comment (santoshmgh) [reply CefSharp/BindingOptions.cs:38]: @amaitland Comment incorporated.

**Comment by santoshmgh** [comment]
@amaitland Kindly let me know if I need to incorporate any other comments.

File: CefSharp/Internals/JavascriptObjectRepository.cs [hunk 11]
```diff
@@ -585,8 +588,18 @@ protected virtual bool TryGetProperty(long objectId, string name, out object res
 
             try
             {
+#if NETCOREAPP
                 result = property.GetValue(obj.Value);
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptObjectRepository.cs:592]: We should be able to exclude the whole method for simplicity.

File: CefSharp/Internals/JavascriptObjectRepository.cs [hunk 12]
```diff
@@ -618,8 +631,18 @@ protected virtual bool TrySetProperty(long objectId, string name, object value,
             }
             try
             {
+#if NETCOREAPP
                 property.SetValue(obj.Value, value);
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptObjectRepository.cs:635]: Same here. We should be able to exclude the whole method for simplicity.

Comment (santoshmgh) [reply CefSharp/Internals/JavascriptObjectRepository.cs]: @amaitland Excluded the TrySetPropperty for .Net core App.

File: CefSharp/Internals/JavascriptObjectRepository.cs (hunk 11 above)
Comment (santoshmgh) [reply CefSharp/Internals/JavascriptObjectRepository.cs]: @amaitland Excluded the TryGetPropperty for .Net core App.

**Comment by santoshmgh** [comment]
@amaitland Kindly let me know how can I run the test case from CefSharp.Test.dll locally. So that I can fix the failing test case.

**Comment by amaitland** [comment]
> Added new QUnit Test cases PropertyInterceptorLogger to CefSharp.Example.ModelBinding

I don't see a qunit test. Can you re-add it please. 

> Kindly let me know how can I run the test case from CefSharp.Test.dll

Tests can be run from within Visual Studio, nothing special required.

**Comment by santoshmgh** [comment]
@amaitland Added the unit test cases to CefSharp.Test.JavascriptBinding.JavaScriptObjectRepositoryFacts

**Comment by amaitland** [comment]
> Added the unit test cases to CefSharp.Test.JavascriptBinding.JavaScriptObjectRepositoryFacts

Please add a separate test case rather than modifying the existing one.

**Comment by santoshmgh** [comment]
@amaitland I have added a separate test case for property Interceptor. Please let me know if you want to update the test name to something else.

**Comment by amaitland** [comment]
Thanks. I'm not too worried about test naming, they can always be improved later without impacting anyone.

**Comment by santoshmgh** [comment]
@amaitland Please confirm if changes for property interceptor will be available as part of cefsharp 98 release.

**Comment by amaitland** [comment]
@santoshmgh If you look at the `Milestone` you'll see this has already been added to `98.1.x`.

**Pull request merged**